sled-agent/src/services.rs

```diff
@@ -893,84 +965,9 @@ impl ServiceManager {
         let omicron_zones_config =
             zones_config.clone().to_omicron_zones_config();
 
-        // Initialize internal DNS only first: we need it to look up the
-        // boundary switch addresses. This dependency is implicit: when we call
-        // `ensure_all_omicron_zones` below, we eventually land in
-        // `opte_ports_needed()`, which for some service types (including Ntp
-        // but _not_ including InternalDns), we perform internal DNS lookups.
-        let all_zones_request = self
-            .ensure_all_omicron_zones(
-                &mut existing_zones,
-                None,
-                omicron_zones_config.clone(),
-                |z: &OmicronZoneConfig| {
-                    matches!(z.zone_type, OmicronZoneType::InternalDns { .. })
-                },
-            )
-            .await?;
-
-        // Initialize NTP services next as they are required for time
-        // synchronization, which is a pre-requisite for the other services. We
-        // keep `OmicronZoneType::InternalDns` because
-        // `ensure_all_omicron_zones` is additive.
-        let all_zones_request = self
-            .ensure_all_omicron_zones(
-                &mut existing_zones,
-                Some(&all_zones_request),
-                omicron_zones_config.clone(),
-                |z: &OmicronZoneConfig| {
-                    matches!(
-                        z.zone_type,
-                        OmicronZoneType::InternalDns { .. }
-                            | OmicronZoneType::BoundaryNtp { .. }
-                            | OmicronZoneType::InternalNtp { .. }
-                    )
-                },
-            )
-            .await?;
-
-        drop(existing_zones);
-
-        info!(&self.inner.log, "Waiting for sled time synchronization");
-
-        retry_notify(
-            retry_policy_local(),
-            || async {
-                match self.timesync_get().await {
-                    Ok(TimeSync { sync: true, .. }) => {
-                        info!(&self.inner.log, "Time is synchronized");
-                        Ok(())
-                    }
-                    Ok(ts) => Err(BackoffError::transient(format!(
-                        "No sync {:?}",
-                        ts
-                    ))),
-                    Err(e) => Err(BackoffError::transient(format!(
-                        "Error checking for time synchronization: {}",
-                        e
-                    ))),
-                }
-            },
-            |error, delay| {
-                warn!(
-                    self.inner.log,
-                    "Time not yet synchronised (retrying in {:?})",
-                    delay;
-                    "error" => ?error
-                );
-            },
-        )
-        .await
-        .expect("Expected an infinite retry loop syncing time");
-
-        let mut existing_zones = self.inner.zones.lock().await;
```

Comment:
This ordering has been removed under the assumption that:

- The NTP zone, which depends on internal DNS in the call to `opte_ports_needed`, will either successfully access *some* internal DNS server (possibly not on this sled!) or fail with a timeout before starting.
- Zones which depend on time being synchronized (which is true for all zones other than NTP and DNS) will fail to start within `start_omicron_zone`, when we check the `time_is_synchronized` boolean.

If these assumptions are upheld, I think this "order-independent" approach to calling `ensure_all_omicron_zones` seems cleaner. It can just be re-called as many times as it needs to be, and zones can keep getting added.